### PR TITLE
88 desktop notification

### LINKED ISSUE #88: Add desktop notifications for when user's pose is bad
# Goal
Include broadcast notification in pipeline
Update pipeline to update pose metric based on time

### PR DESCRIPTION
**Pull request by C-Begley**
# Details

- When the user is in a bad posture for a certain period of time, the pipeline will send a desktop notification to the user
- Notification currently sends if users' posture is bad for more than 10 seconds and at a maximum every 2 minutes _(this perhaps can be added as a value in settings)_

To test the code:
-Make the Broadcast receiver on a Linux device (can also be the Raspberry Pi):
```
cd src/notifications
cmake build .
make
cd ../../
./src/notifications/NotificationReceiver
```
-Run posture perfection on the Raspberry Pi
```
./scripts/build.sh
./PosturePerfection
```

# Screenshots/Gif (if appropriate)
Screenshot from 2021-04-07 15-31-07
Screenshot from 2021-04-07 15-32-34
Screenshot from 2021-04-07 15-33-01

# Checklist

## Code
- [ ] I have added some tests
- [x] I have run the tests locally to ensure they all pass by running `.scripts/build.sh -enable-testing`

## Documentation
- [ ] I have updated documentation appropriately (GitHub Pages)
- [ ] I have tested the updated GitHub Pages locally to ensure the changes to documentation render correctly (see this [Guidance](https://ese-peasy.github.io/PosturePerfection/guidance.html) page)

Closes #88

### CODE REVIEW

**Comment by miklasr** [comment]
I get the following error (for PosturePerfection) when attempting to test with a bad posture:
```
terminate called after throwing an instance of 'char const*'
```

**Comment by miklasr** [comment]
As we discussed we should consider moving this to a separate repo now. At the very least splitting it into a separate directory on this repo.

**Comment by C-Begley** [comment]
> I get the following error (for PosturePerfection) when attempting to test with a bad posture:
> 
> ```
> terminate called after throwing an instance of 'char const*'
> ```

This seems to be a raspberry pi specific error, I'll investigate

**Comment by C-Begley** [comment]
> As we discussed we should consider moving this to a separate repo now. At the very least splitting it into a separate directory on this repo.

I think perhaps we could merge this request and we can look at that in another pull request?

**Comment by Andrew-Ritchie** [comment]
The notifications are really cool, good job! Although when I run the server and application on the Pi it looks like the new detected posture is not refreshing properly. 

Screenshot 2021-04-08 at 11 01 40

**Review by Andrew-Ritchie — APPROVED** [review]
This is working really well now, good job. The notifications are a really cool and useful feature!

Screenshot 2021-04-08 at 12 59 37

File: src/posture_estimator.cpp [hunk 1]
```diff
@@ -24,17 +24,18 @@
...
+#define NOTIFICATION_TIME 30000    ///< 100 seconds
+                          
 
 namespace PostureEstimating {
 
 std::string stringJoint(Joint joint) {
   switch (joint) {
     case Head:
-      return "head";
+      return "h./ead";
```
Comment (Ashwin-MJ) [inline src/posture_estimator.cpp]: Typo?

**Review by Ashwin-MJ — COMMENTED** [review]
Not entirely sure what the desired behaviour is, this is what happened for me when starting `NotificationReceiver` and then starting `PosturePerfection` :
1. Initial notification to say that `PosturePerfection` has started (Works correctly every time)
2. I set an ideal pose and while I remain in a good posture no notifications are sent (Seems correct to me)
3. When I adopt a bad posture for 10 seconds I get the alert letting me know I have an imperfect posture" (Also seems correct)

Not entirely sure what should happen after this - I switched back to a good posture and then to a bad one, and no notifications were delivered. Also, I never got the notification for "Undefined" at any point

**Comment by C-Begley** [comment]
> Not entirely sure what the desired behaviour is, this is what happened for me when starting `NotificationReceiver` and then starting `PosturePerfection` :
> 
>     1. Initial notification to say that `PosturePerfection` has started (Works correctly every time)
> 
>     2. I set an ideal pose and while I remain in a good posture no notifications are sent (Seems correct to me)
> 
>     3. When I adopt a bad posture for 10 seconds I get the alert letting me know I have an imperfect posture" (Also seems correct)
> 
> 
> Not entirely sure what should happen after this - I switched back to a good posture and then to a bad one, and no notifications were delivered. Also, I never got the notification for "Undefined" at any point

Yeah, currently it takes 10 mins for another notification to happen. If you want to turn it down, just change NOTIFICATION_TIME to something smaller like 30000 ms. There is a plan to add this to the settings page I believe

File: src/posture_estimator.cpp (hunk 1 above)
Comment (C-Begley) [reply src/posture_estimator.cpp]: Whoops

**Comment by Ashwin-MJ** [comment]
Could we have two separate timers for `Undefined` and `Bad`? The logic as follows:
- If I am `Undefined` for 5 minutes I get a single notification "Are you there?". I then don't get another notification like this for at least 10 minutes (again, only if I am `Undefined` for 5 minutes)
- If I am in a `Bad` posture for 10 seconds I get a single notification "Imperfect posture...". I then don't get another notification like this for at least 2/3 minutes (again, only if I am in a `Bad` posture for 10 seconds)

**Comment by C-Begley** [comment]
> Could we have two separate timers for `Undefined` and `Bad`? The logic as follows:
> 
>     * If I am `Undefined` for 5 minutes I get a single notification "Are you there?". I then don't get another notification like this for at least 10 minutes (again, only if I am `Undefined` for 5 minutes)
> 
>     * If I am in a `Bad` posture for 10 seconds I get a single notification "Imperfect posture...". I then don't get another notification like this for at least 2/3 minutes (again, only if I am in a `Bad` posture for 10 seconds)

Updated to use these timings. You may want to turn them down for testing

File: src/posture_estimator.h [hunk 2]
```diff
@@ -105,6 +107,123 @@ struct PoseStatus {
...
+      false;  ///< Boolen which indicates if timer is running (True = running)
+  /**
+   * @brief Constructor for `PostureEstimating::MessageTimer`
+   * @param notificationTimers List of `PostureEstimating::DelayTimers` that
+  must not be running for message to be broadcast
+  * @param broadcaster  `Notify::NotifyBroadcast` used to broadcast the message
+  * @param msg message to be sent
+  * @param time The time to wait
+  before broadcasting message
+  */
```
Comment (miklasr) [inline src/posture_estimator.h]: Can you rerun the formatter to realign this?

**Review by Ashwin-MJ — APPROVED** [review]

**Review by miklasr — APPROVED** [review]
Superb work! I've tested with RPi as the sender and desktop as receiver of notifications.

**Pull request merged**